packages/apollo-server-core/src/determineApolloConfig.ts

```diff
@@ -0,0 +1,109 @@
+// to warn if old environment variables are used.
+export function determineApolloConfig(
+  input: ApolloConfigInput | undefined,
+  // For backwards compatibility.
+  // AS3: Drop support for deprecated 'engine'.
+  engine: EngineReportingOptions<any> | boolean | undefined,
+  logger: Logger,
+): ApolloConfig {
+  if (input && engine !== undefined) {
+    // There's a more helpful error in the actual ApolloServer constructor.
+    throw Error('Cannot pass both `apollo` and `engine`');
+  }
+  const apolloConfig: ApolloConfig = { graphVariant: 'current' };
+
+  const {
+    APOLLO_KEY,
+    APOLLO_GRAPH_ID,
+    APOLLO_GRAPH_VARIANT,
+    // AS3: Drop support for deprecated `ENGINE_API_KEY` and `ENGINE_SCHEMA_TAG`.
+    ENGINE_API_KEY,
+    ENGINE_SCHEMA_TAG,
+  } = process.env;
+
+  // Determine key.
+  if (input?.key) {
+    apolloConfig.key = input.key;
+  } else if (typeof engine === 'object' && engine.apiKey) {
+    apolloConfig.key = engine.apiKey;
+  } else if (APOLLO_KEY) {
+    if (ENGINE_API_KEY) {
+      logger.warn(
+        'Using `APOLLO_KEY` since `ENGINE_API_KEY` (deprecated) is also set in the environment.',
+      );
+    }
+    apolloConfig.key = APOLLO_KEY;
+  } else if (ENGINE_API_KEY) {
+    logger.warn(
+      '[deprecated] The `ENGINE_API_KEY` environment variable has been renamed to `APOLLO_KEY`.',
+    );
+    apolloConfig.key = ENGINE_API_KEY;
+  }
+
+  // Determine key hash.
+  if (apolloConfig.key) {
+    apolloConfig.keyHash = createSHA('sha512')
+      .update(apolloConfig.key)
+      .digest('hex');
+  }
+
+  // Determine graph id.
+  if (input?.graphId) {
+    apolloConfig.graphId = input.graphId;
+  } else if (APOLLO_GRAPH_ID) {
+    apolloConfig.graphId = APOLLO_GRAPH_ID;
+  } else if (apolloConfig.key) {
+    // This is the common case: if the given key is a graph token (starts with 'service:'),
+    // then use the service name written in the key.
+    const parts = apolloConfig.key.split(':', 2);
+    if (parts[0] === 'service') {
+      apolloConfig.graphId = parts[1];
+    }
+  }
+
+  // Determine variant.
+  if (input?.graphVariant) {
+    apolloConfig.graphVariant = input.graphVariant;
+  } else if (typeof engine === 'object' && engine.graphVariant) {
+    if (engine.schemaTag) {
+      throw new Error(
+        'Cannot set more than one of apollo.graphVariant, ' +
+          'engine.graphVariant, and engine.schemaTag. Please use apollo.graphVariant.',
+      );
+    }
+    apolloConfig.graphVariant = engine.graphVariant;
+  } else if (typeof engine === 'object' && engine.schemaTag) {
+    // No need to warn here, because ApolloServer's constructor should warn about
+    // the existence of `engine` at all.
+    apolloConfig.graphVariant = engine.schemaTag;
+  } else if (APOLLO_GRAPH_VARIANT) {
+    if (ENGINE_SCHEMA_TAG) {
+      throw new Error(
+        '`APOLLO_GRAPH_VARIANT` and `ENGINE_SCHEMA_TAG` (deprecated) environment variables must not both be set.',
+      );
```

Comment:
Do we want to warn and carry on gracefully in these `throw` cases? Or are we throwing because if the values are different we don't know which is the right one to pick?

I know this complicates the logic a bit but we could carry on if:
1) they're identical (safe)
2) we just decide that newer config option takes more precedence (arguably less safe)

Reply:
This throw is preserving existing behavior. Has anyone complained about it yet?

In general I think "make sure untouched configuration still works" is a good goal but "let people write confusing config that is half old and half new" isn't positive.